mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -263,17 +288,21 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
     }
 
     val w = if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0) else col($(weightCol))
-    val instances: RDD[Instance] =
-      dataset.select(col($(labelCol)), w, col($(featuresCol))).rdd.map {
-        case Row(label: Double, weight: Double, features: Vector) =>
-          Instance(label, weight, features)
+    val off = if (!isDefined(offsetCol) || $(offsetCol).isEmpty) lit(0.0) else col($(offsetCol))
+    val instances: RDD[OffsetInstance] =
+      dataset.select(col($(labelCol)), w, off, col($(featuresCol))).rdd.map {
+        case Row(label: Double, weight: Double, offset: Double, features: Vector) =>
+          OffsetInstance(label, weight, offset, features)
       }
 
     val model = if (familyObj == Gaussian && linkObj == Identity) {
       // TODO: Make standardizeFeatures and standardizeLabel configurable.
+      val wlsInstances: RDD[Instance] = instances.map { instance =>
+        Instance(instance.label - instance.offset, instance.weight, instance.features)
```

Comment:
another thing you can do, if you don't want to change the hierarchy, is to move the initialization of instances inside the if/else.  Then, for weightedleastsquares, you can just create RDD[Instance], but for IRLS weighted instances.